src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -109,20 +109,127 @@ static CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE spvcrossType2E_TYPE
 // returns true if successfully added all the info to self, false if incompatible with what's already in our pipeline or incomplete (e.g. missing spec constants)
 NBL_API2 bool CSPIRVIntrospector::CPipelineIntrospectionData::merge(const CSPIRVIntrospector::CStageIntrospectionData* stageData, const ICPUShader::SSpecInfoBase::spec_constant_map_t* specConstants)
 {
-    return false;
+
+    // merge descriptors
+    for (uint32_t i = 0u; i < ICPUPipelineLayout::DESCRIPTOR_SET_COUNT; i++)
+    {
+        std::ostringstream debug;
+        debug << "ds ID: " << i << std::endl;
+        const auto& bindingInfos = stageData->getDescriptorSetInfo(i);
```

Comment:
you need to be able to put `*this` back in its original state BEFORE the `merge` call if you encounter a failure.

So I'd take a full copy of the descriptor infos BEFORE entering the loop, then `std::move` the copies back into the members if you manage to make it to the other side without returning false.

Reply:
Theoretically to avoid such copies,  you could split it into two loops, one that validates, and another that sets, but that would require you already know that the `stageData` is internally consistent, meaning there are no two bindings in there that would fail validation against themselves (i.e. no case of binding B being declared both as a UBO and an SSBO)